service/history/workflow/workflow_task_state_machine.go

```diff
@@ -465,7 +477,7 @@ func (m *workflowTaskStateMachine) AddWorkflowTaskCompletedEvent(
 
 	// Capture if WorkflowTaskScheduled and WorkflowTaskStarted events were created
 	// before calling m.beforeAddWorkflowTaskCompletedEvent() because it will delete workflow task info from mutable state.
-	workflowTaskScheduledStartedEventsCreated := !m.ms.IsTransientWorkflowTask()
+	workflowTaskScheduledStartedEventsCreated := !m.ms.IsTransientWorkflowTask() && workflowTask.Type != enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE
```

Comment:
Not necessary for this PR but this might be a good thing to encapsulate into executionInfo